crates/common/tedge_config_macros/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+#[doc(inline)]
+pub use tedge_config_macros_macro::define_tedge_config;
+
+extern crate self as tedge_config_macros;
+
+use camino::Utf8PathBuf;
+pub use connect_url::*;
+use default::*;
+use doku_aliases::*;
+pub use option::*;
+use std::net::IpAddr;
+use std::net::Ipv4Addr;
+use std::num::NonZeroU16;
+use std::path::PathBuf;
+
+mod connect_url;
+mod default;
+mod doku_aliases;
+mod option;
+
+static DEFAULT_ROOT_CERT_PATH: &str = "/etc/ssl/certs";
+
+define_tedge_config! {
```

Comment:
Just to confirm what I understand:

* The `define_tedge_config!` defines a `TEdgeConfigDto` struct which hierarchy reflects the macro call structure.
* This DTO manages read / write / types / docs / examples as well as default values.
* The DTO structure is to be used to update the config but one needs to call `TEdgeCsonfigReader::from_dto(&dto, &TEdgeConfigLocation)` to get a `TEdgeConfig` where all the defaults have been resolved (or the errors raised).
* The `TEdgeConfig` struct has the same hierarchy and be used as a regular struct with no method invocations to get specific settings however deep there are in the hierarchy.

Is this correct? If so this is really appealing!

Reply:
That is essentially correct, and it's encouraging that it could be understood (even if only after following discussions on the other PR)!

I think at the moment I'm generating the documentation from the reader, so read-only configurations are treated in the same way there to read-write configurations, but that doesn't make a huge difference to generating them from the DTO (and we may just want to get rid of read-only configurations entirely).

The intention with the reader is to only expose an immutable borrow to other crates so they can only read the underlying configuration. I intend to enable writing to the configuration, e.g. to set `c8y.url` in `tedge connect c8y` in a similar way to the current setup, where we give the caller `&mut TEdgeConfigDto` and they just apply the necessary updates.